Updating Key Vault project with bug fixes, refactoring and updating JWK and renaming test project folder.

pomortaz opened the pull request:
The JsonWebKey related classes are moved to a new project to accomodate modularity of the projects and references. More functionalities and tests for JsonWebKey is ported in. 
There are additional field and bug fixes for KeyVault SDK.
Additional tests are added for crypto library.
Tests are moved to Microsoft.Azure.\* folders to be aligned with the main projects and namespaces.

herveyw-msft committed:
Align folder, project and namespaces names.

pomortaz committed:
Adding KV managed field, pattern validation, fixed K for JsonWebKey, bug fix, add permission enum types documentation

pomortaz committed:
Moving KV recorded tests to their renamed directory namespace.

pomortaz committed:
Refactor webkey to a new package

herveyw-msft committed:
Update RsaKeyTests for uniformity of tests with Java and cross platform signature verification.

pomortaz committed:
Moving in JWK constructor for RSA and additional functionalities and tests, adding documentations.

pomortaz committed:
Update the credential scanner path for KeyVault test after the namespace is changed

pomortaz commented:
@herveyw, @rohmano and @kunalc please review.

kunalcha commented:
>         public const string Pfx = "application/x-pkcs12";

Consider adding application/x-pem-file as well as a supported content type.

---
Refers to: src/KeyVault/Microsoft.Azure.KeyVault/Customized/CertificateBundle.cs:70 in 8b65176. [](commit_id = 8b65176e8a6f6fe1bad29b2c5840171e44004cae, deletion_comment = False)

kunalcha commented:
>     public static class WellKnownIssuers

I won't recommend this type on client side as it will cause friction with new issuer providers that get added on service side. I'd recommend specifying the list via documentation.

---
Refers to: src/KeyVault/Microsoft.Azure.KeyVault/Customized/CertificateBundle.cs:78 in 8b65176. [](commit_id = 8b65176e8a6f6fe1bad29b2c5840171e44004cae, deletion_comment = False)

kunalcha commented on src/KeyVault/Microsoft.Azure.KeyVault/Customized/KeyVaultClientExtensions.cs:
```diff
@@ -413,18 +413,19 @@ public static partial class KeyVaultClientExtensions
         /// <param name="cancellationToken">Optional cancellation token</param>
         /// <returns>The updated certificate.</returns>
-        public static async Task<CertificateBundle> UpdateCertificateAsync(this IKeyVaultClient operations, string certificateIdentifier, CertificateAttributes certificateAttributes = null, IDictionary<string, string> tags = null, CancellationToken cancellationToken = default(CancellationToken))
+        public static async Task<CertificateBundle> UpdateCertificateAsync(this IKeyVaultClient operations, string certificateIdentifier, CertificatePolicy certificatePolicy = default(CertificatePolicy), CertificateAttributes certificateAttributes = null, IDictionary<string, string> tags = null, CancellationToken cancellationToken = default(CancellationToken))
```
>UpdateCertificateAsync [](start = 52, length = 22)

Is there a similar method to just update the certificate policy. Certificate Policy is a child resource and can be updated by itself.

pomortaz replied:
Yes.

kunalcha commented on src/KeyVault/Microsoft.Azure.KeyVault/Customized/ObjectIdentifier.cs:
```diff
@@ -253,8 +335,16 @@ public CertificateOperationIdentifier(string identifier)
+    /// <summary>
+    /// The Key Vault issuer identifier.
+    /// </summary>
     public sealed class IssuerIdentifier : ObjectIdentifier
```
>IssuerIdentifier [](start = 24, length = 16)

I will recommend naming this as CertificateIssuerIdentifier as it's a part of certificates collection.

kunalcha commented on src/KeyVault/Microsoft.Azure.KeyVault/Generated/Models/CertificateIssuerUpdateParameters.cs:
```diff
@@ -0,0 +1,97 @@
+            }
+            if (this.Provider != null)
+            {
+                if (this.Provider.Length > 20)
```
>if (this.Provider.Length > 20) [](start = 16, length = 30)

Do you think these checks are required on client side?

pomortaz replied:
Yes. They are generated from the swagger spec.

kunalcha commented on src/KeyVault/Microsoft.Azure.KeyVault/Generated/Models/KeyBundle.cs:
```diff
@@ -59,5 +63,11 @@ public KeyBundle(JsonWebKey key = default(JsonWebKey), KeyAttributes attributes
         public IDictionary<string, string> Tags { get; set; }
 
+        /// <summary>
+        /// Gets or sets true if the secret's lifetime is managed by key vault.
```
>/// Gets or sets true if the secret's lifetime is managed by key vault. [](start = 8, length = 71)

Consider mentioning an example. i.e. if this is a key backing a certificate, then managed will be true.

kunalcha commented on src/KeyVault/Microsoft.Azure.KeyVault/Generated/Models/KeyBundle.cs:
```diff
@@ -59,5 +63,11 @@ public KeyBundle(JsonWebKey key = default(JsonWebKey), KeyAttributes attributes
+        /// Gets or sets true if the secret's lifetime is managed by key vault.
+        /// </summary>
+        [JsonProperty(PropertyName = "managed")]
+        public bool? Managed { get; set; }
```
>set; [](start = 36, length = 4)

set should not be allowed.

kunalcha commented on src/KeyVault/Microsoft.Azure.KeyVault/Generated/Models/SecretBundle.cs:
```diff
@@ -83,5 +86,11 @@ public SecretBundle(string value = default(string), string id = default(string),
+        /// Gets or sets true if the secret's lifetime is managed by key vault.
+        /// </summary>
+        [JsonProperty(PropertyName = "managed")]
+        public bool? Managed { get; set; }
```
>set [](start = 36, length = 3)

set should not be allowed (public).

kunalcha commented:
>         /// Gets or sets the key id for certificate.

Consider mentioning that if this is a secret backing a KV certificate, then this field specified the corresponding key backing the KV certificate. The field is read only (not settable). In other cases this value is null.

---
Refers to: src/KeyVault/Microsoft.Azure.KeyVault/Generated/Models/SecretBundle.cs:84 in 8b65176. [](commit_id = 8b65176e8a6f6fe1bad29b2c5840171e44004cae, deletion_comment = False)

pomortaz commented:
@kunalcha, about the WellKnownIssuers, it is defined as a static class with string types as values. So it is not fixed like enum and other values can be used without any need to update the SDK.

pomortaz committed:
making managed readonly, changed IssuerIdentifier to CertificateIssuerIdentifier and added content type.

markcowl commented:
@kunalcha, please indicate signong off on the changed PR.

markcowl commented on src/KeyVault/Microsoft.Azure.KeyVault.Cryptography.Tests/Tests/Cryptography/Algorithms/RsaKeyTests.cs:
```diff
@@ -0,0 +1,156 @@
+﻿//
```
WHy apache instead of MIT license?

pomortaz replied:
I'll change it to MIT license. And will change the rest of the files.

markcowl commented on src/KeyVault/Microsoft.Azure.KeyVault.WebKey/JsonWebKey.cs:
```diff
@@ -0,0 +1,637 @@
+    /// <summary>
+    /// As of http://tools.ietf.org/html/draft-ietf-jose-json-web-key-18
+    /// </summary>
+    public class JsonWebKey
```
This looks like a generated model.  If it is, it should go in the Generated folder

herveyw replied:
Some probably generated but lots of additions and type is generally usable outside the SDK and we don't want the SDK dependencies. See discussion on Slack re: ms-client-external and the need to have references to externally defined types for a better experience.

pomortaz replied:
@markcowl everything that does not have a header that says it is generated, is not generated.

pomortaz committed:
Change KV Apache license to MIT license and minor changes to fields:
Make kid readonly.
Remove the restriction on the length of the provider.
Fix key permission constant captalization.
Remove predefined values for well known issuers.

kunalcha commented on src/KeyVault/Microsoft.Azure.KeyVault/Generated/Models/SecretBundle.cs:
```diff
@@ -78,10 +84,19 @@ public SecretBundle(string value = default(string), string id = default(string),
 
         /// <summary>
-        /// Gets or sets the key id for certificate.
+        /// Gets or sets if this is a secret backing a KV certificate, then
```
>or sets [](start = 17, length = 7)

This set should be private too.

pomortaz replied:
Thanks. Change kid to readonly.

kunalcha commented:
:shipit:

pomortaz commented:
@markcowl can you please merge?
